src/entry.rs

```diff
@@ -597,24 +598,51 @@ impl<'a> EntryFields<'a> {
         return Ok(());
 
         #[cfg(any(unix, target_os = "redox"))]
-        fn set_perms(dst: &Path, mode: u32, preserve: bool) -> io::Result<()> {
+        fn set_perms(
+            dst: &Path,
+            f: Option<&mut std::fs::File>,
+            mode: u32,
+            preserve: bool,
+        ) -> io::Result<()> {
             use std::os::unix::prelude::*;
 
             let mode = if preserve { mode } else { mode & 0o777 };
-
             let perm = fs::Permissions::from_mode(mode as _);
-            fs::set_permissions(dst, perm)
+            match f {
+                Some(f) => f.set_permissions(perm),
+                None => fs::set_permissions(dst, perm)
+            }
         }
+
         #[cfg(windows)]
-        fn set_perms(dst: &Path, mode: u32, _preserve: bool) -> io::Result<()> {
-            let mut perm = r#try!(fs::metadata(dst)).permissions();
-            perm.set_readonly(mode & 0o200 != 0o200);
-            fs::set_permissions(dst, perm)
+        fn set_perms(
+            dst: &Path,
+            f: Option<&mut std::fs::File>,
+            mode: u32,
+            _preserve: bool,
+        ) -> io::Result<()> {
+            match f {
+                Some(f) => {
+                    let mut perm = f.metadata()?.permissions();
+                    perm.set_readonly(mode & 0o200 != 0o200);
+                    f.set_permissions(perm)
+                }
+                None => {
+                    let mut perm = r#try!(fs::metadata(dst)).permissions();
```

Comment:
FWIW if this is run across it's fine to replace all instances of `r#try!` with `?`